packages/x-internals/src/warning/slotWarning.ts

```diff
@@ -0,0 +1,19 @@
+import { warnOnce } from './warning';
+
+export function checkSlot(slotName: string, slot: Function) {
+  // If the render function is passed as anonymous arrow function, its name will be the same as the slot name:
+  //   slots={{ toolbar: () => <div /> }} // slots.toolbar.name === 'toolbar'
+  // If a React component is passed, it's usually capitalized:
+  //   slots={{ toolbar: Toolbar }} // slots.toolbar.name === 'Toolbar'
+  // The rare edge cases are:
+  //   slots={{ toolbar: toolbar }} // slots.toolbar.name === 'toolbar' - false positive. It's a component, not a render function
+  //   slots={{ toolbar: function Toolbar() {} }} // slots.toolbar.name === 'Toolbar' - false negative. It's a render function, not a component
```

Comment:
I suspect that the right solution would be https://github.com/mui/material-ui/issues/31096 here.

Reply:
Do you think it's the right solution? I don't expect a lot of people to install the plugin.
Warnings on the other hand would work for all users.